### PR TITLE
update build system & add Limine build pipeline

### PR DESCRIPTION
**Pull request by riblanc**
This pull request introduces a new build pipeline. It allows the use of the [Limine bootloader](https://github.com/limine-bootloader/limine) instead of GRUB by setting the `BOOTLOADER` variable in the Makefile to `BOOTLOADER=limine`.

It also introduces a few changes to the basic pipeline (327fc94):
- fix fclean rule to clean .zig-cache instead of zig-cache folder
- add a theme_clean rule in Themes.mk
- use of the theme_clean rule instead of a system command to remove theme files in thtmes.zig (avoiding an error when "themes" folder doesn't exist)
- replace kfs.iso gitignore rule with *.iso
- remove every steps declaration from the BuildContext structure. This started to get very complicated, as it was hard to see which step depended on which ones, where the steps were defined, and so on.

Using limine allows us to build our final iso without system-dependent files, as grub could do.
I thought it would be interesting to use it in our github actions: It's no longer necessary to build a docker image, run it, etc., which is rather slow. So I reworked the CI part to use limine.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,51 +1,61 @@
-BUILD_ARGS ?= --summary all --verbose
-ZIG ?= zig
+BUILD_ARGS ?= --summary all --verbose -Dbootloader=$(BOOTLOADER)
+
+# get the last bootloader option
+BOOTLOADER ?= $(shell echo $$(ls ".bootloader-"* 2>/dev/null || echo grub) | cut -d'-' -f2)
```
Comment (jgiron42) [inline Makefile:4]: shouldn't the BOOTLOADER var be defined before being used for BUILD_ARGS?

File: build.zig [hunk 2]
```diff
@@ -4,55 +4,55 @@ const Builder = @import("std").Build;
...
 
-    @import("build/disk_image.zig").install_iso_folder(&context);
-    @import("build/kernel.zig").build_executable(&context);
-    @import("build/themes.zig").install_themes(&context);
-    @import("build/disk_image.zig").build_disk_image(&context);
-    @import("build/qemu.zig").add_step_run(&context);
+    // Dependencies
+    kernel.step.dependOn(&themes.step);
+    b.getInstallStep().dependOn(&install_disk_image.step);
+    b.getInstallStep().dependOn(&install_disk_image.step);
```
Comment (jgiron42) [inline build.zig:56]: seems like a mistake (I'm not sure if you wanted another dependency or if you just accidentally duplicated the line)

File: build/Makefiles/Zig.mk [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+	@TERM=xterm tput setaf 3
+	@echo "WARNING: Zig is not installed. Trying to use a local version (v$(ZIG_VERSION))."
+	@TERM=xterm tput sgr0
+
+.PHONY: install_zig
+install_zig: $(ZIG_LOCAL) zig_warning
+
+.PHONY: uninstall_zig
+uninstall_zig:
+	rm -rf zig-linux-x86_64-$(ZIG_VERSION)
```
Comment (jgiron42) [inline build/Makefiles/Zig.mk:29]:
```suggestion
	rm -rf $(ZIG_LOCAL)
```

File: Makefile (hunk 1 above)
Comment (riblanc) [reply Makefile:4]: I don't think the order is really important here.

https://www.gnu.org/software/make/manual/html_node/Reading-Makefiles.html

> GNU make does its work in two distinct phases. During the first phase it reads all the makefiles, included makefiles, etc. and internalizes all the variables and their values and implicit and explicit rules, and builds a dependency graph of all the targets and their prerequisites. During the second phase, make uses this internalized data to determine which targets need to be updated and run the recipes necessary to update them

As you can read in the mentioned page, the '?=' operator is deferred to the second phase.

But if you want, I can move it to `BOOTLOADER` for greater clarity.

Comment (riblanc) [reply Makefile]: I made the changes

File: build.zig (hunk 2 above)
Comment (riblanc) [reply build.zig]: Done

File: build/Makefiles/Zig.mk (hunk 3 above)
Comment (riblanc) [reply build/Makefiles/Zig.mk]: Done

**Review by jgiron42 — APPROVED** [review]

**Pull request merged**